[FEAT] MembershipCouponMenu 기능 구현 - 문주영

juyeongMoon888 opened the pull request:
# 목표
카페에서 주문 후 결제하면 프로모션을 적용하고 적립을 제공한다. 

# 기능
### 설계 상세 내용
https://www.notion.so/ohgiraffers/277649136c11803fb7afc1a61be7e2b5

# 📚 학습 내용
- 메뉴판 출력 시 하드코딩을 제거하고, Enum 순회로 유지보수성을 개선했습니다. 
- inCafe 플래그로 “입장만 한 손님”과 “주문 손님” 경계를 분리하여 SOLID의 SRP 원칙을 활용했습니다.

# ✅ 체크리스트
## 메뉴 관리 플로우대로 구현하였는가 ?
- 동일 손님의 다음 주문에서 쿠폰/적립금 사용을 지원해야 하나, 아직 미구현입니다.
- 쿠폰 및 적립금의 차기 주문(동일 고객) 사용 기능은 추가 예정입니다.
- 현재는 인메모리 저장으로 구현했고, 추후 DB 연동으로 전환할 예정입니다.
## 예상 가능한 오류가 없는가 ?
- Gap: 예상 가능한 오류(엣지 케이스) 사전 점검 부족
- Cause: 기능 구현·기본 동작 테스트에 집중
- Next: 입력 검증/상태 전이/중복 처리 등 실패 시나리오 테스트 보강


# 🤔 질문사항 (선택)
- 책임 분리를 어떻게하면 숨쉬듯이 할 수 있을지..

juyeongMoon888 committed:
feat: Menu enum 도입

juyeongMoon888 committed:
refacfor: Menu -> MenuType으로 변경

juyeongMoon888 committed:
feat: service 계층 추가

juyeongMoon888 committed:
feat: GuestService 추가

juyeongMoon888 committed:
feat: Guest 엔티티 추가

juyeongMoon888 committed:
feat: Order가 게스트 식별자(guestId) 필드 추가 및 생성자 연계

juyeongMoon888 committed:
feat: OrderService 추가

juyeongMoon888 committed:
feat: 아이템별로 고정 ID 추가

juyeongMoon888 committed:
refactor: MenuType -> MenuItem으로 수정

juyeongMoon888 committed:
feat: OrderItem 추가

juyeongMoon888 committed:
feat: 입력 수집 메서드(readMenuSelection, readQuantity) 추가

juyeongMoon888 committed:
feat: 메뉴 스냅샷(이름, 단가)를 캡처해 필드에 고정시키는 captureMenuSnapshot() 추가

juyeongMoon888 committed:
feat: Order에 id/guestId/아이템 컬렉션 및 합계 계산 로직 추가

juyeongMoon888 committed:
feat: Order에 totalCups필드 및 연산 추가

juyeongMoon888 committed:
refactor: currentGuest를 루프 밖에서 1회 생성하여 방문 세션 동안 유지

juyeongMoon888 committed:
feat: GuestRepository 추가 - guestMap, findById(), save()생성

juyeongMoon888 committed:
feat: findGuest 메서드 추가

juyeongMoon888 committed:
feat: OrderRepository 추가

juyeongMoon888 committed:
feat: 게스트와 주문 생성, 주문 조회 및 저장 메서드 추가

juyeongMoon888 committed:
refactor: createOrder 반환을 inline Variable로 단순화

juyeongMoon888 committed:
refactor: 저장소 단일화 및 의존 주입 정리
- 앱 시작 시 GuestRepository/OrderRepository 각각 1회만 생성
- 공유 인스턴스 주입으로 일관된 Map 사용 보장
- startOrderForGuest/ findById 동작 안정화 (guestMap null 해결)

juyeongMoon888 committed:
fix: values 순회 중 null id로 인한 NPE 제거 → key 기반 조회(Map.get)로 변경

juyeongMoon888 committed:
refactor: 엔티티 seq 필드 제거, id 생성 책임을 Repository로 일원화

- 엔티티는 식별자 보관만 수행(불변성 강화)
- Repository 단일 경로에서 id 발급•관리
- 객체 내부 seq 제거로 id 직접 수정/충돌 가능성 차단

juyeongMoon888 committed:
feat: 프로모션 미충족(5잔 미만)일 때 기본 결제 완료 안내 추가

juyeongMoon888 committed:
feat: 포인트 잔액 필드 추가

juyeongMoon888 committed:
feat: OrderStatus(enum) 추가

juyeongMoon888 committed:
refactor: findById를 key 기반 조회로 변경

juyeongMoon888 committed:
feat: 결제 완료 주문에 대해 포인트 적립 메서드 추가

juyeongMoon888 committed:
feat: OrderStatus 필드 추가

juyeongMoon888 committed:
MenuType -> MenuItem 파일명 변경

JSeHoone commented:
### 개인적인 의견
멤버십/쿠폰 관리 파트에서 주문관리까지 생각하셔서 구현하고자 하는 기능들이 더 커진 것 처럼 보입니다 !
> 주문의 총 합계의 10%가 적립되는 것 때문에 주문관리까지 생각하신게 아닌가 싶습니다

이미 회원가입을 한 사람들이거나 회원 가입을 하겠다고 한 사람들을 대상으로 하는 기능들을 구현했으면 구현하고자 하는 기능이 조금 줄지 않았을까 싶습니다 !

아래는 주문 등록을 모두 뺀 새로운 멤버를 생성하는 기능을 생각해본 것 입니다 !
1. 회원 등록
```java
public void registerMember(int totalAccount, String phoneNumber) {
    // 적립금 계산 로직
    int membershipPoint = totalAccount * 0.1;

    // 회원 생성로직 (return 멤버ID)
    Member newMember = new Member() // 주영님 코드에서는 Guest로 생각하면 될 것 같습니다 !
    newMember.setPhoneNumber(phoneNumber);
    newMember.setPoint(membershipPoint);

    // 회원 저장하기
    memberRepository.save(newMember);
}
```

### 질문사항에 대한 의견

책임 분리를 어떻게 숨쉬듯이 분리할 수 있을지 저도 너누 궁금하고 공감됩니다 🥲
다 같이 공부하고 의견을 나누다보면 쪼개는 연습이 되지 않을까요 ?!
화이팅입니다 🔥

dawwson commented on src/main/java/org/BackOffice/services/loyalty/MembershipCouponMenu.java:
```diff
@@ -1,4 +1,158 @@
+            }
+        }
+        return id;
+    }
```
`사용자 입력 메뉴 번호` -> `MenuItem id`로 변환하는 과정이 하드코딩 되어 있어, 메뉴 추가 시 이 부분도 수정 가능성이 있어보입니다.
도메인 안으로 들어간다면 좀 더 응집도 있게 작성될 수 있을 것 같습니다.

```
// MenuItem.java (예시)
public enum MenuItem {
    AMERICANO(1001, "아메리카노", 4500),
    // ...

    public static MenuItem findByDisplayIndex(int index) {
        // index (1, 2, 3, ...) 에 해당하는 MenuItem을 찾아 반환
        // 예: MenuItem.values()[index - 1] (bounds check 필요)
    }
}
// 이렇게 하면 새로운 메뉴가 추가되어도 이 매핑 로직은 수정할 필요가 없습니다.
```

juyeongMoon888 replied:
감사합니다! 이대로 수정해보겠습니다. 메서드를 이렇게 줄일 수도 있군요!😀

dawwson reviewed (approved):
검토할 PR이 많아서 Gemini의 도움을 받았습니다..😅 고생하셨습니다.!

[좋은 점]
- DB 도입을 고려하여 domain, repository, service로 분리하신 점
- 핵심 도메인 엔티티를 잘 정의하신 것 같습니다.

[아쉬운 점]
- 세부적인 책임 분리를 개선하면 더 좋을 것 같습니다.

juyeongMoon888 committed:
feat: switchToPaid() 메서드 추가 - Order의 주문 상태 변경

juyeongMoon888 committed:
feat: LoyaltyRepository 추가 - 손님별 적립금 관리

juyeongMoon888 committed:
feat: LoyaltyAccount 엔티티 추가 (적립금)

juyeongMoon888 committed:
feat: LoyaltyService 생성

juyeongMoon888 committed:
fix: save() 수정 - 불필요한 매개변수 제거

juyeongMoon888 committed:
feat: confirmOrderIntent() 추가 - 주문의사 확정 메서드 추가

juyeongMoon888 committed:
refactor(input): Scanner를 static 공유 인스턴스로 변경하여 중복 생성 방지

juyeongMoon888 committed:
feat: AppContext 추가 - Repository 생성 역할 책임

juyeongMoon888 committed:
refactor(factory): ServiceFactory 무상태화 및 서비스 생성 로직 분리

- Guest/Order/Loyalty 서비스 생성을 전담하는 팩토리 도입
- AppContext에서 Repository 제공, MembershipCouponMenu에서 팩토리 통해 메서드 생성
- 서비스 생성자 주입으로 의존성 연결(불변)

juyeongMoon888 committed:
refactor(factory): MembershipCouponMenu에 생성자 주입 도입(Service/Repository)
- 내부 new 제거, AppContext/Factory에서 조립한 의존성 주입

juyeongMoon888 committed:
refactor: 손님 선택을 플로우에서 분리

- selectOrCreateGuest/GuestService.selectExistingGuest 추가로 손님 식별 단계 분리
- startOrder()은 주문 시작 책임만 수행
- SRP 강화
- startOrder가 guestI 인자를 요구하도록 호출부 업데이트

juyeongMoon888 committed:
refactor: 게스트 생성/주문 결합 메서드 제거 및 createOrder(guestId)로 단순화

- startOrderForGuest, createOrder(currentGuestId) 제거
- 주문 생성 책임을 createOrder(guestId) 단일 메서드로 집중
- SRP 강화
- Guest 상태 갱신은 상위 플로우로 이관

juyeongMoon888 committed:
refactor: 게스트/주문 분리로 인한 불필요한 의존성 제거

juyeongMoon888 committed:
refactor: 메뉴판 번호 → ID 변환 로직을 MenuItem로 응집하고 switch 제거

- menuDisplayIndexToId 메서드 제거
- readMenuSelection()에서 메뉴판 번호→ID 변환까지 처리해 메서드 수 축소

- MenuItem.findByDisplayIndex(int) 도입: enum 순회로 디스플레이 번호를 도메인 ID로 해석
- 매칭 실패 시 -1 반환하도록 계약 명시
- UI 경계에서 해석(디스플레이 번호→도메인 ID) 책임을 일원화해 응집도 향상

juyeongMoon888 committed:
refactor: 진행중 주문 할당 책임을 분리하고 assignOpenOrderToGuest 도입

 - OrderService가 수행하던 “게스트에 오픈 주문 연결” 책임을 게스트 측 서비스로 이전
 - assignOpenOrderToGuest(guestId, orderId) 추가로 책임 명확화(SRP)
 - 호출부 일괄 수정

juyeongMoon888 committed:
feat: LoyaltyAccount 생성 시 결제금액의 10%를 적립하도록 로직 추가

juyeongMoon888 committed:
refactor: save를 create/update로 분리하고 ID 할당을 생성 시점으로 한정
 - 업데이트 시 불필요한 ID 변경 및 시퀀스 증가 방지
 - 호출부 수정

juyeongMoon888 committed:
refactor: Guest에서 pointBalance 제거

juyeongMoon888 committed:
refactor: LoyaltyAccount 생성 단순화 및 적립 로직 분리
- 생성자에서 결제금액 기반 적립 제거 → guestId만 설정, 초기 balance=0
- 잔액 조회 게터 추가: getBalance()
- 적립 로직을 도메인 메서드로 분리: setBalance(totalPay) (결제금액의 10% 누적)

juyeongMoon888 committed:
refactor: AppContext에 Repository 게터 추가로 접근 경로 표준화

juyeongMoon888 committed:
style: 불필요한 주석 제거

juyeongMoon888 committed:
feat: getOrCreateForGuest(guestId) 추가로 계정 존재 보장
- guestId로 조회 후 없으면 balance=0인 LoyaltyAccount 생성·저장 뒤 반환(절대 null 아님)
- 불필요한 Guest import 제거, 레포지토리 책임 명확화

juyeongMoon888 committed:
feat: LoyaltyService에 DI 도입 및 적립 흐름 단순화
- LoyaltyRepository 생성자 주입으로 의존성 명시
- accruePoints 시그니처 변경: 레포지토리 파라미터 제거 → 내부 필드 사용
- 계정 조회 로직을 getOrCreateForGuest(guestId)로 일원화
- 적립 처리 account.setBalance(totalPay)로 누적 방식 명확화
- 불필요한 GuestService 의존 제거
- 불필요한 Guest, GuestRepository import 제거

juyeongMoon888 committed:
refactor: ServiceFactory에서 순수 팩토리 메서드로 단순화
- AppContext/자기참조 제거 및 불필요한 상태 의존성 제거
- 각 create*Service가 생성한 서비스 인스턴스 반환하도록 변경

juyeongMoon888 committed:
feat: 포인트 적립 흐름 제거하고 '포인트 사용 적용'으로 전환
- accrueLoyaltyPointsForPaidOrder(orderId) 제거
- applyPointsUse(int points, int orderId) 도입: 주문로드 → 포인트 사용 반영 → 저장 → 사용량 반환
- 호출부 전면 교체 및 사용 로그 내용 변경(콘솔 출력)

juyeongMoon888 committed:
feat: MenuAdapter 추가 - InMemoryData 대응

juyeongMoon888 committed:
feat: startCallOrder 추가 - InMemoryData (사전입력) 주문 경로 대응

- 인메모리 기반 주문 플로우를 startCallOrder로 분리
- *전화 주문에 비유해서 메서드명에 'Call' 삽입
- 수동 입력(startOrder)과 채널 분리하여 라우팅 명확화
- 결제 단계는 공용 proceedToCheckout 재사용
- 중복 로직 제거 및 입력 채널별 책임 SRP 강화

juyeongMoon888 committed:
feat: 입력 채널 InputChannel 도입 (MANUAL, PRESET) 및 사용처 일괄 변경
- Method -> InputChannel로 명칭 변경
- VOICE -> PRESET로 의미 정정(사전 입력/인메모리 채널)
- MembershipCouponMenu 포함 전 호출부 스위치/라우팅 업데이트

juyeongMoon888 committed:
refactor(domain,repo,service): 식별자 타입 int → Long 일괄 전환 (orderId-guestId)

- InMemoryData 스키마(Long)와 호환을 위해 Order.id, Order.guestId, Guest.id 등을 Long으로 변경
- 레포지토리/맵 키, 시퀀스 생성기 및 메서드 시그니처·반환타입 전부 Long 기반으로 수정

juyeongMoon888 committed:
feat: Method → InputChannel로 명칭 변경

juyeongMoon888 committed:
refactor: save를 create/update로 분리하고 ID 할당을 생성 시점으로 한정
 - 업데이트 시 불필요한 ID 변경 및 시퀀스 증가 방지
 - 호출부 수정

juyeongMoon888 committed:
refactor: saveOrder제거하고 생성/갱신 경로 분리 (createOrder/updateOrder)
- createOrder → orderRepo.create 사용으로 ID 부여 단일화
- updateOrder 신설 → orderRepo.update로 내용 갱신 시 ID 불변 보장
- 중복 책임이던 saveOrder 삭제로 API 명확화

juyeongMoon888 committed:
feat(format): PointsFormatter 추가 - 포인트 표시 포맷 전용 컴포넌트
- 접두/접미(예: "+", "P") 등 불변 설정 기반 포맷 수행
- 입력 포인트를 문자열로 변환(부호/천단위 구분 지원)
- 표시 책임 분리

juyeongMoon888 committed:
feat: MembershipCouponMenu에 PointsFormatter 도입해 멤버십 요약 출력 포맷 통일 - 포인트
- MembershipCouponMenu에서 PointsFormatter 인스턴스 생성
- printMembershipSummary()에서 포인트 표시를 포맷터로 위임(부호/천단위 일관화)

Commit:
feat: CouponType enum 추가

Commit:
feat: Coupon 엔티티 추가

Commit:
feat: CounponStatus(ISSUED, REDEEMED, EXPIRED, CANCELLED) - 쿠폰 상태 추가
